src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,177 @@
+            : this(environment, unhandledExceptionHandler, consoleBufferTimeout: DefaultConsoleBufferTimeout, autoStart: true)
+        {
+        }
+
+        internal ConsoleWriter(IEnvironment environment, Action<Exception> exceptionHandler, TimeSpan consoleBufferTimeout, bool autoStart)
+        {
+            bool consoleEnabled = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingDisabled) != "1";
+
+            if (consoleEnabled)
+            {
+                // We are going to used stdout, but do we write directly or use a buffer?
+                _consoleBuffer = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferSize) switch
+                {
+                    "-1" => Channel.CreateUnbounded<string>(),              // buffer size of -1 indicates that buffer should be enabled but unbounded
+                    "0" => null,                                            // buffer size of 0 indicates that buffer should be disabled
+                    var s when int.TryParse(s, out int i) && i > 0 => Channel.CreateBounded<string>(i),
+                    _ => Channel.CreateBounded<string>(DefaultBufferSize),  // default behavior is to use buffer with default size
+                };
+
+                if (_consoleBuffer == null)
+                {
+                    _writeEvent = Console.WriteLine;
+                }
+                else
+                {
+                    _writeEvent = WriteToConsoleBuffer;
+                    _consoleBufferTimeout = consoleBufferTimeout;
+                    if (autoStart)
+                    {
+                        bool batched = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferBatched) switch
+                        {
+                            "0" => false,     // disable batching by setting to 0
+                            _ => true,        // default behavior is batched
+                        };
+
+                        _consoleBufferReadLoop = ProcessConsoleBufferAsync(batched);
+                    }
+                }
+            }
+            else
+            {
+                _writeEvent = (string s) => { };
+            }
+
+            _exceptionhandler = exceptionHandler;
+        }
+
+        public void WriteHandler(string evt)
+        {
+            _writeEvent(evt);
+        }
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            try
+            {
+                while (_consoleBuffer.Writer.TryWrite(evt) == false)
+                {
+                    // Buffer is currently full, wait until writing is permitted.
+                    // This is the downside of using channels, we are on a sync code path and so we have to block on this task
+                    var writeTask = _consoleBuffer.Writer.WaitToWriteAsync().AsTask();
+                    if (writeTask.WaitAsync(_consoleBufferTimeout).Result == false)
```

Comment:
This sync over async (.Result) concerns me. Did you originally have this doing a spin wait to avoid this? Because the way the code is now it's not clear to me why there's a while loop here if it only attempts once.

Reply:
It attempts more than once. Its a while loop because after calling WaitToWriteAsync() we still have to call TryWrite, and TryWrite can still fail again if we raced with another writer and they got there first. The only situation we give up here is if the timeout expires (and therefore WaitAsync returns false).

I am not sure how to reconcile your concern about the .Result with your earlier comment about ordering. If we decide to enforce ordering, then this thread has to be blocked.. I don't think we want to make all of our logging codepaths async.

Probably worth mentioning that I am bumping up the buffer size based on Nick's suggestions, which decreases the frequency of these cases.

Let me know if you have alternatives in mind that you want me to consider.

Reply:
One option that I forgot to mention is that we could immediately write to the console if the buffer is full. This would remove the WaitToWriteAsync call completely, so no blocking on Task.Result. It would increase the potential for out-of-order logs, but that might be an acceptable tradeoff.

Reply:
I am not 100% sure how `Console` implementation works in C#, but is writing a point of contention? If the buffer is full, and this starts writing directly, would that slow down the ability to even drain the buffer since now more things are fighting for `Console`?

Reply:
Yes, I think that outcome is theoretically possible, but how it plays out depends a bit on the nature of the situation that causes the buffer to be full in the first place. For example, threadpool starvation can cause delays on the buffer consumer side, in which case Console.Write contention is not the problem.

The PR has been updated with a new synchronization approach that still follows the rule of "try to wait some time to maximize chance of preserving log ordering" but avoids the pitfalls of blocking on a `GetAwaiter().GetResult()`.